### DISCUSSION TITLE
Release's download URL is redirecting

### DISCUSSION THREAD

**Discussion by TheAspen** [community/community#46420] [Repositories] [answered]
Hi,

I'm trying to add my application to Microsoft store and I must provide the download link to my .exe -file.
This .exe -file can be downloaded from my public Github repository's releases. However, when I try to provide this link to Microsoft store
I get the error notification of "Provided URL redirects to another URL, please provide the original URL.". So the Github download URl seems to redirect somewhere. 
The URL itself is working as it suppose to, you are able to download that .exe file with it.

I provided the url as the documentation said (https://docs.github.com/en/repositories/releasing-projects-on-github/linking-to-releases) like this: /owner/name/releases/latest/download/application.exe

So why Github link is redirecting? Where it redirect and how I can get this "original" link which I suppose to provide to Microsoft store?
Should I even use Github repository as a content delivery network (CDN)?

_- TheAspen_

**Comment by Lamparter** [comment]
**Hi @TheAspen!**
I think I have the answer to your question.
###### 😰 Please bear in mind this _is_ slightly complex, however..

You can find the _original CDN server link_ (the redirectless download) by taking the following steps:
1. Go to your GitHub `releases` page (`https://github.com/owner/repo/releases` and find the latest release
![image](https://user-images.githubusercontent.com/71598437/216818364-c2d0c7b2-ab30-4a2d-83a5-a36543eb3d23.png)
2. Download the latest release package (in your case, the `exe` file.)
![image](https://user-images.githubusercontent.com/71598437/216818421-b720dd61-09ec-4957-89e9-9db58d837d63.png)
3. Go to your browser's downloads page: e.g. `edge://downloads` or `chrome://downloads`
![image](https://user-images.githubusercontent.com/71598437/216818527-67044abd-751a-4995-9493-09d01567543c.png)
4. Find your package file (something like `AppName.exe`)
![image](https://user-images.githubusercontent.com/71598437/216818559-72bc28da-5983-4ecd-a7d5-91a78c98abe0.png)
5. You'll see a link attached to it (probably beginning with `https://objects.githubusercontent.com` or `https://codeload.github.com`). This is your static download URL.

**Reply by TheAspen** [reply]
Hey,

This actually worked, thanks! 
Though I had to use Edge -browser, did not work via Chrome which is kind of weird.

**Reply by TheAspen** [reply]
@Lamparter 

Hi,

Bad news, today I noticed that the old link which I used 2 days ago, is not working anymore. So I look into the download URL and noticed that link is not quite the same anymore. Signature and timestamp have changed and probably will change again in the future. 

Microsoft store publish system accepted that link, but I think it will become inoperative at some point. So I'm not able to proceed with the publish process.

**Reply by Lamparter** [reply]
> @Lamparter
> 
> Hi,
> 
> Bad news, today I noticed that the old link which I used 2 days ago, is not working anymore. So I look into the download URL and noticed that link is not quite the same anymore. Signature and timestamp have changed and probably will change again in the future.
> 
> Microsoft store publish system accepted that link, but I think it will become inoperative at some point. So I'm not able to proceed with the publish process.

There _is_ a way to fix this, but it's unconventional.
I'll post a new answer.. one second..

**Reply by Lamparter** [reply]
> @Lamparter
> 
> Hi,
> 
> Bad news, today I noticed that the old link which I used 2 days ago, is not working anymore. So I look into the download URL and noticed that link is not quite the same anymore. Signature and timestamp have changed and probably will change again in the future.
> 
> Microsoft store publish system accepted that link, but I think it will become inoperative at some point. So I'm not able to proceed with the publish process.

@TheAspen can I clarify something: is there a problem with the link you generated from the instructions (it expired)? Or do you mean that that link does not give you the latest version, instead a fixed version?

**Reply by TheAspen** [reply]
@Lamparter The link is not working at all. When I try to use the link, I got the http-error: "This page isn't working. HTTP-ERROR 401". And now it is also happening with the link I generated yesterday, so the links seems to expire.

**Reply by Lamparter** [reply]
> @Lamparter The link is not working at all. When I try to use the link, I got the http-error: "This page isn't working. HTTP-ERROR 401". And now it is also happening with the link I generated yesterday, so the links seems to expire.

Hmm.. 401 is a login/authorisation error..

**Reply by TheAspen** [reply]
Yes, which is very odd because the repository and the release are public. I can link it here also: [Release](https://github.com/TheAspen/Dice_Roller_2000/releases/tag/v1.0.0)
So there should not be any logins or so when you download the .exe file from that release.

But something is happening to the links when time past, because the same link won't work if you wait couple of days or so.

**Comment by Belphemur** [comment]
I'm on the same path, the only way to do this, is to actually use a website that would use github API to find the wanted release and then stream the file directly, not redirect.

This is stupid, but that the only way if you don't want to host the file yourself manually somewhere for each release.

**Comment by Lamparter** [comment, marked as answer]
@TheAspen There is one way.. but it's kinda unconventional. However, this one is guaranteed to work.
1. Create a new branch in your repo called `release`
![image](https://user-images.githubusercontent.com/71598437/218172399-1ac0386e-7d22-41f7-bfc4-993d92162ad2.png)
2. Switch to your new branch
![image](https://user-images.githubusercontent.com/71598437/218172507-3275cb6d-6bfe-4a32-89c2-10ba070f5022.png)
3. Navigate to 'Upload a file'
![image](https://user-images.githubusercontent.com/71598437/218172620-ff36f5f4-41db-4644-bd1f-b9c810601861.png)
4. Find your project's installer object file
![image](https://user-images.githubusercontent.com/71598437/218173674-10784138-02c3-44e5-9b76-771df8f34358.png)
5. Generate your release download URL. e.g: `https://raw.githubusercontent.com/RiversideValley/TestRepo/release/MyApp_0.5.0.appx`
![image](https://user-images.githubusercontent.com/71598437/218174651-e0bafa87-50ab-49ac-9103-bd1ac590a35c.png)

Your release download URL is completely permanent and non-redirecting.

**Reply by TheAspen** [reply]
@Lamparter  I will mark this as an answer. However I already went to Azure and use it as my CDN. 
I might try this at some point though. Maybe at that point when I publish an update or new version of the application. 

I would like to use only Github for this hobby project, I like the simplicity it provides. :)

Thanks for the all effort, I really appreciate it!

**Reply by JuamBer** [reply]
It don't work for me because the .exe binary is too big. The only thing left for me is to create a website to upload the binaries there

**Reply by Lamparter** [reply]
> However I already went to Azure and use it as my CDN.

@JuamBer you can use Azure DevOPS or another free Azure service to publish the binary.
###### Also, mustn't that be a **whopping** big executable to have a binary file that large!

**Reply by JuamBer** [reply]
I've never worked with Azure, but I'll keep an eye on it. Thank you so much

**Comment by Yahwehloves** [comment]
https://github.com/Yahwehloves/People-.gitgit remote add origin https://github.com/Yahwehloves/People-.git
git branch -M main
git push -u origin main

**Reply by Lamparter** [reply]
Hi @Yahwehloves, I believe you have posted this into the wrong place...
